File: app/src/main/java/com/afollestad/nocknock/util/MathUtil.java
package com.afollestad.nocknock.util;

import android.graphics.Path;
import android.support.design.widget.FloatingActionButton;
import android.view.View;

/**
 * @author Aidan Follestad (afollestad)
 */
public final class MathUtil {

    public static Path bezierCurve(FloatingActionButton fab, View rootView) {
        final int fabCenterX = (int) (fab.getX() + fab.getMeasuredWidth() / 2);
        final int fabCenterY = (int) (fab.getY() + fab.getMeasuredHeight() / 2);

        final int endCenterX = (rootView.getMeasuredWidth() / 2) - (fab.getMeasuredWidth() / 2);
        final int endCenterY = (rootView.getMeasuredHeight() / 2) - (fab.getMeasuredHeight() / 2);

        final int halfX = (fabCenterX - endCenterX) / 2;
        final int halfY = (fabCenterY - endCenterY) / 2;
        int mControlX = endCenterX + halfX;
        int mControlY = endCenterY + halfY;
        mControlY -= halfY;
        mControlX += halfX;

        Path path = new Path();
        path.moveTo(fab.getX(), fab.getY());
        path.quadTo(mControlX, mControlY, endCenterX, endCenterY);

        return path;
    }
}

File: utilities/src/main/java/com/afollestad/nocknock/utilities/util/MathUtil.kt
/*
 * Licensed under Apache-2.0
 *
 * Designed and developed by Aidan Follestad (@afollestad)
 */
package com.afollestad.nocknock.utilities.util

import android.graphics.Path
import android.view.View

/** @author Aidan Follestad (afollestad) */
object MathUtil {

  fun bezierCurve(
    targetView: View,
    rootView: View
  ): Path {
    val fabCenterX = (targetView.x + targetView.measuredWidth / 2).toInt()
    val fabCenterY = (targetView.y + targetView.measuredHeight / 2).toInt()

    val endCenterX = rootView.measuredWidth / 2 - targetView.measuredWidth / 2
    val endCenterY = rootView.measuredHeight / 2 - targetView.measuredHeight / 2

    val halfX = (fabCenterX - endCenterX) / 2
    val halfY = (fabCenterY - endCenterY) / 2

    var controlX = endCenterX + halfX
    var controlY = endCenterY + halfY

    controlY -= halfY
    controlX += halfX

    val path = Path()
    path.moveTo(targetView.x, targetView.y)
    path.quadTo(
        controlX.toFloat(),
        controlY.toFloat(),
        endCenterX.toFloat(),
        endCenterY.toFloat()
    )

    return path
  }
}
